### PR TITLE
[Tabular] GPU optimization, scheduling for parallel_local fitting strategy

### PR DESCRIPTION
**Pull request by prateekdesai04**
*Description of changes:*
Distribute load all folds on all available GPUs equally for simplicity, this is used when `fold_fitting_strategy` is set to `parallel_local`.

We notice that Ray's scheduler sometimes places unequal load on all GPUs, i.e., when `num_gpus = 3` and `num_bag_folds = 17` then by default all GPUs get 3 tasks each but for the remaining two tasks ray sometimes schedules them on one GPU for example GPU 0 instead of giving GPU 0 and GPU 1 the two tasks, causing uneven load.

This change addresses this issue by doing exactly what is mentioned above.

How to test:

1. Fit a small tabular model with `num_bag_folds` and `num_gpus` to any value
2. Set model hyperparameters to `ag.ens.fold_fitting_strategy':'parallel_local`
3. Simultaneously run `watch -n 0.1 nvidia-smi` on a new window
4. See ray process spin up evenly on GPUs 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 1]
```diff
@@ -726,10 +748,23 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
...
 
+    def _calculate_gpu_assignment(self, gpu_assignments: Dict, task_id: int, gpus_per_task: int, total_gpus: int):
+        if gpus_per_task >= 1:
+            gpu_id = task_id * gpus_per_task
+            assigned_gpus = []
+            for i in range(gpus_per_task):
+                assigned_gpus.append(gpu_id + i % total_gpus)
+            gpu_assignments[task_id] = assigned_gpus
+        else:
+            gpu_id = task_id % self.num_gpus
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: ensure consistent usage of either `total_gpus` or `self.num_gpus`

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+        time_start=time.time(),
+        time_limit=None,
+        models=[],
+        oof_pred_proba=np.array([]),
+        oof_pred_model_repeats=np.array([]),
+        save_folds=True,
+        num_cpus=num_cpus,
+        num_gpus=num_gpus,
+    )
+    return ParallelLocalFoldFittingStrategy(**args)
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py:134]: Lets actually make some of the more generic parameters here have defaults in the actual class so we don't need to specify so many things here in the test explicitly.

For example, `X_pseudo, y_pseudo, `sample_weight, time_start, time_limit, save_folds` can all be given default values equal to what you have here.

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+    """
+    Create a Ray remote function that reports GPU information.
+
+    This function is created outside the test class so it can be serialized
+    and sent to Ray remote workers.
+    """
+    try:
+        import ray
+    except ImportError:
+        return None
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: Why are we allowing None / ImportError? Feels like this doesn't help us with tests.

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+        gpu_assignments = {}
+
+        gpu_assignments = strategy._calculate_gpu_assignment(
+            gpu_assignments=gpu_assignments,
+            task_id=0,
+            gpus_per_task=0,
+            total_gpus=0
+        )
+
+        assert gpu_assignments[0] == []
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: we should verify more than just this from the output, such as len(gpu_assignments).

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+            gpu_assignments = strategy._calculate_gpu_assignment(
+                gpu_assignments=gpu_assignments,
+                task_id=task_id,
+                gpus_per_task=0.5,
+                total_gpus=1
+            )
+
+        # Both tasks should be assigned to GPU 0
+        assert gpu_assignments[0] == [0]
+        assert gpu_assignments[1] == [0]
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: verify there are only 2 tasks

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+
+        expected = [0, 1, 0, 1]
+        for task_id, expected_gpu in enumerate(expected):
+            assert gpu_assignments[task_id] == [expected_gpu]
+
+        # Submit tasks and collect reports
+        task_refs = []
+        for task_id in range(4):
+            ref = ray_task_report_gpu_info.remote(task_id, gpu_assignments[task_id])
+            task_refs.append(ref)
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: can we verify that a task waits to be scheduled until its GPU is available?

can we check the available GPUs as jobs finish one by one?

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+            gpu_assignments = strategy._calculate_gpu_assignment(
+                gpu_assignments=gpu_assignments,
+                task_id=task_id,
+                gpus_per_task=0.5,
+                total_gpus=2
+            )
+
+        expected = [0, 1, 0, 1]
+        for task_id, expected_gpu in enumerate(expected):
+            assert gpu_assignments[task_id] == [expected_gpu]
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: Need to test extra scenario where we have 4 tasks, minimum 1 GPU per task, and only 2 GPUs available. In this scenario:

If task 2 finishes before task 1, shouldn't task 3 launch with GPU 1 instead of task 4 launching with GPU 1?

Doing it this way means the GPU used for a task isn't pre-determined before the task is ran, but rather is dynamically allocated based on available resources.

This seems like it would be most efficient.

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+        ray.shutdown()
+
+    @pytest.mark.multi_gpu
+    def test_ray_gpu_assignment_multiple_gpus_round_robin_task_execution(self):
+        """
+        Integration Test: Ray tasks verify round-robin GPU assignment
+        This test should run on continuous_integration_multigpu.yaml
+
+        Scenario: 2 GPUs, 4 tasks
+        num_gpus_per_task = 2 / 4 = 0.5 (fractional, round-robin)
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: Test edge cases, ensure logic works as expected

- 2 GPUs, 3 tasks, per-task = 2/3 GPU
- 2 GPUs, 3 tasks, per-task = 1 GPU
- 3 GPUs, 2 tasks, per-task = 1.5 GPUs (assert raises?)

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 9]
```diff
@@ -726,10 +752,27 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
...
+        if gpus_per_task >= 1:
+            gpu_id = task_id * gpus_per_task
+            assigned_gpus = []
+            for i in range(gpus_per_task):
+                assigned_gpus.append(gpu_id + i % total_gpus)
+            gpu_assignments[task_id] = assigned_gpus
+        else:
+            gpu_id = task_id % total_gpus
+            gpu_assignments[task_id] = [gpu_id]
+        return gpu_assignments
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: Why are we doing in-place edits to gpu_assignments?
Why not just return the value we want to set it to, and then

```
gpu_assignments[task_id] = self._calculate_gpu_assignment(task_id, gpus_per_task, total_gpus)
```

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 10]
```diff
@@ -726,10 +752,27 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
...
+    def _calculate_gpu_assignment(self, gpu_assignments: Dict, task_id: int, gpus_per_task: int, total_gpus: int):
+        if total_gpus == 0:
+            logger.debug(f"No GPUs available, CPU-only mode for task {task_id}")
+            gpu_assignments[task_id] = []
+            return gpu_assignments   
+        if gpus_per_task >= 1:
+            gpu_id = task_id * gpus_per_task
+            assigned_gpus = []
+            for i in range(gpus_per_task):
+                assigned_gpus.append(gpu_id + i % total_gpus)
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: Is this a bug?

Shouldn't it be:

```
assigned_gpus.append((gpu_id + i) % total_gpus)
```

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 11]
```diff
@@ -726,10 +752,27 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
...
+            gpu_assignments[task_id] = assigned_gpus
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: Should we sort the GPU IDs?

`assigned_gpus = sorted(assigned_gpus)`

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 12]
```diff
@@ -726,10 +752,27 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
 
         self._update_bagged_ensemble_times()
 
+    def _calculate_gpu_assignment(self, gpu_assignments: Dict, task_id: int, gpus_per_task: int, total_gpus: int):
+        if total_gpus == 0:
+            logger.debug(f"No GPUs available, CPU-only mode for task {task_id}")
+            gpu_assignments[task_id] = []
+            return gpu_assignments   
+        if gpus_per_task >= 1:
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py:765]: Should we add an assert that it is not a float if >=1?

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 13]
```diff
@@ -726,10 +752,27 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
 
         self._update_bagged_ensemble_times()
 
+    def _calculate_gpu_assignment(self, gpu_assignments: Dict, task_id: int, gpus_per_task: int, total_gpus: int):
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: gpus_per_task: int | float

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 14]
```diff
@@ -726,10 +752,27 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
...
+        if total_gpus == 0:
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py:762]: assert total_gpus >= 0
assert gpus_per_task >= 0
assert task_id >= 0

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 15]
```diff
@@ -417,6 +419,14 @@ def _ray_fit(
     model_sync_path: Optional[str] = None,
 ):
     import ray  # ray must be present
+    import torch
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: What if torch isn't installed?

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py [hunk 16]
```diff
@@ -0,0 +1,490 @@
...
+
+        # Verify each task sees its assigned GPU
+        for result in results:
+            task_id = result['task_id']
+            expected_gpu = task_id
+            assert result['CUDA_VISIBLE_DEVICES'] == str(expected_gpu)
+            # Each task should see exactly 1 GPU
+            if result['torch_gpu_count'] != 'torch not available':
+                assert result['torch_gpu_count'] == 1, \
+                    f"Task {task_id} expected 1 GPU, saw {result['torch_gpu_count']}"
```
Comment (Innixma) [inline tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: I notice a lot of code dupe between these tests. Try to re-use logic into functions as much as possible. I think we can make this code a lot shorter by re-using verification logic.

**Review by Innixma — COMMENTED** [review]
Added first round of comments, awesome contribution!

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py (hunk 3 above)
Comment (prateekdesai04) [reply tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: removed import to fail if `ray` is not installed or detected

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py (hunk 7 above)
Comment (prateekdesai04) [reply tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: Will look at dynamic allocation in a follow-up PR as it seems to be a little more complex than this approach.
But added more tests where available GPUs = n, tasks = m, and gpus_per_task = n where (n > m)

File: tabular/tests/unittests/resource_allocation/test_gpu_assignment.py (hunk 16 above)
Comment (prateekdesai04) [reply tabular/tests/unittests/resource_allocation/test_gpu_assignment.py]: converted verification logic into separate functions

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 17]
```diff
@@ -417,6 +419,11 @@ def _ray_fit(
     model_sync_path: Optional[str] = None,
 ):
     import ray  # ray must be present
+    gpu_ids = assignments.get(task_id, [])
+    if gpu_ids:
+        # Set CUDA_VISIBLE_DEVICES to the assigned GPU IDs
+        os.environ['CUDA_VISIBLE_DEVICES'] = ','.join(map(str, gpu_ids))
+        logger.debug(f"Set CUDA_VISIBLE_DEVICES to {gpu_ids}")
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: What if gpu_ids is `[]`?
This will treat `if gpu_ids` as False, which means we won't actually edit `CUDA_VISIBLE_DEVICES`.
Would it perhaps be better to do `gpu_ids = assignments.get(task_id, None)` and `if gpu_ids is not None:`?

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 18]
```diff
@@ -408,7 +408,9 @@ def _ray_fit(
     y: Union[str, pd.DataFrame],
     X_pseudo: Union[str, pd.DataFrame],
     y_pseudo: Union[str, pd.DataFrame],
+    task_id: int,
     fold_ctx: Dict[str, Any],
+    assignments: Dict[int, List[int]],
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: `assignments` is too generic of a name. Consider something like `task_to_gpu_ids_map`.

Also, why are we sending in a dict when this is a fit call for a specific task_id.
Why not just pass in `gpu_ids` directly?

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 19]
```diff
@@ -661,9 +679,10 @@ def _update_bagged_ensemble_child_resources(self):
     def _run_parallel(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_limit_fold, head_node_id):
         job_refs = []
         job_fold_map = {}
+        gpu_assignments = {}
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py:684]: Why are we passing an empty dictionary?
Why not have `gpu_assignments` be a variable of the class: `self.gpu_assignments`?

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 20]
```diff
@@ -705,7 +726,9 @@ def _run_pseudo_sequential(self, X, y, X_pseudo, y_pseudo, model_base_ref, time_
         a job could start fitting a model while the results are processed; resulting in the fit running out of memory
         due to the overhead of processing and storing the result.
         """
-        for job in self.jobs:
+        gpu_assignments = {}
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py:731]: Why are we passing an empty dictionary?
Why not have `gpu_assignments` be a variable of the class: `self.gpu_assignments`?

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 21]
```diff
@@ -807,7 +855,9 @@ def _fit(
             y=y_ref,
             X_pseudo=X_pseudo_ref,
             y_pseudo=y_pseudo_ref,
+            task_id=task_id,
             fold_ctx=fold_ctx_ref,
+            assignments=gpu_assignments,
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: Don't pass a task more information than it needs. It doesn't need to know about other task's GPU assignments.

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py (hunk 18 above)
Comment (prateekdesai04) [reply core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: Changed the variable name.
Sending in the whole dict may help us achieve 2 things
1. If we have to use the whole dict for dynamic allocation logic in the future
2. Monitor the state of the dict on all ray workers (better for logging and debugging)

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py (hunk 17 above)
Comment (prateekdesai04) [reply core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: `_calculate_gpu_assignment()` returns `[]` in CPU only mode hence the current implementation works, would replacing `[]` with `None` make any difference ?

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py (hunk 19 above)
Comment (prateekdesai04) [reply core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py:684]: My assumption/understanding may be wrong, but what if the ensemble starts training another model before it finishes training the first model or basically another instance of FoldFittingStrategy is called in the same process?

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py (hunk 20 above)
Comment (prateekdesai04) [reply core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py:731]: ditto as above

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py (hunk 21 above)
Comment (prateekdesai04) [reply core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: same comment as to why we pass the dict - for dynamic allocation and logging/monitoring in the future for debugging states across multiple ray processes

Comment (Innixma) [reply core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: A task doesn't need to know the state of other processes. And it isn't necessary for dynamic allocation, that is handled in the main process, not the workers.

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 22]
```diff
@@ -425,6 +432,17 @@ def _ray_fit(
...
+    try:
+        import torch
+        visible_gpus = os.environ.get("CUDA_VISIBLE_DEVICES", "not set")
+        num_gpus = torch.cuda.device_count()
+        current_gpu = torch.cuda.current_device() if torch.cuda.is_available() else "N/A"
+        print(f"[GPU DEBUG] CUDA_VISIBLE_DEVICES={visible_gpus}, Torch sees {num_gpus} GPUs, Using GPU {current_gpu}", flush=True)
+    except ImportError:
+        pass
+    except Exception as e:
+        print(f"[GPU DEBUG] Could not get GPU info: {e}", flush=True)
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py]: This is too heavy of a debug statement, we don't want to import torch unless we have to. You can add a debug flag for this, but don't import torch by default in workers or it slows everything down.
